### PR TITLE
fix(participants): move interval update of talking time counter to the store

### LINKED ISSUE #11501: Talking time is transferred to other participants
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Be in call with someone speaking
2. Check "talking time"

### Expected behaviour
Timer should reflect correct individual value

### Actual behaviour
Some of the speaking signals were transferred to wrong participants (See screenshot for  user 3, 4)
image

Might be a regression of #11254

## Talk app

**Talk app version:** 18

### PR DESCRIPTION
**Pull request by Antreesy**
### ☑️ Resolves

* Fix #11501
	* Data in the store are correct and accessible by token + attendeeId
	* Because of virtual scroller in the participant list, interval update in the component can't be used, as it attached to the Vue component object, not the participant data
	* Moved interval update to the store, so each second we update a value there
	* Component rely only on store update and re-rendered once per second (as before), but with correct value

### 🖼️ Screenshots / Screencasts

🏚️ Before | 🏡 After
-- | --
Reproducible | Non-reproducible

### 🚧 Tasks

- [ ] speaking don't rely on participantStore, so may be moved to `callExtras` store :pineapple: 
  - [ ] Check if SpeakingStatusHandler class supports Pinia

### 🏁 Checklist

- [x] 🌏 Tested with Chrome, Firefox and Safari or should not be risky to browser differences
- [x] 🖥️ Tested with Desktop client or should not be risky for it 
- [ ] ⛑️ Tests are included or not possible

### CODE REVIEW

File: src/components/RightSidebar/Participants/Participant.vue [hunk 1]
```diff
@@ -864,22 +862,20 @@ export default {
...
-				clearInterval(this.speakingInterval)
-				this.speakingInterval = null
+
+		timeSpeaking() {
+			if (!this.participantSpeakingInformation) {
+				return null
+			}
+			if (this.isParticipantSpeaking === undefined) {
+				return 0
 			}
```
Comment (DorraJaouad) [inline src/components/RightSidebar/Participants/Participant.vue:872]:
```suggestion
			if (!this.participantSpeakingInformation || !this.isParticipantSpeaking) {
				return null
```

No need to differentiate, boolean value of `timeSpeaking` will be the same

File: src/store/participantsStore.js [hunk 2]
```diff
@@ -414,7 +416,7 @@ const mutations = {
 	 * @param {string} data.attendeeId - the attendee ID of the participant in conversation.
 	 * @param {boolean} data.speaking - whether the participant is speaking or not
 	 */
-	setSpeaking(state, { token, attendeeId, speaking }) {
+	setSpeakingTime(state, { token, attendeeId, speaking }) {
```
Comment (DorraJaouad) [inline src/store/participantsStore.js:419]: I suggest to keep the name and apply the following

```js
	setSpeaking(state, { token, attendeeId, speaking }) {
		// create a dummy object for current call
		if (!state.speaking[token]) {
			Vue.set(state.speaking, token, {})
		}

		if (!state.speaking[token][attendeeId]) {
			Vue.set(state.speaking[token], attendeeId, { speaking, lastTimestamp: 0, totalCountedTime: 0 })
			return
		}
			
		state.speaking[token][attendeeId].speaking = speaking
	},
```

File: src/store/participantsStore.js [hunk 3]
```diff
@@ -433,9 +435,56 @@ const mutations = {
...
+		// create a dummy object for current call
+		if (!state.speakingInterval[token]) {
+			Vue.set(state.speakingInterval, token, {})
+		}
+		if (!state.speakingInterval[token][attendeeId]) {
+			Vue.set(state.speakingInterval[token], attendeeId, interval)
+		} else if (!interval) {
+			clearInterval(state.speakingInterval[token][attendeeId])
+			state.speakingInterval[token][attendeeId] = interval
+		}
```
Comment (DorraJaouad) [inline src/store/participantsStore.js:462]:
```suggestion
		if (!interval) {
			clearInterval(state.speakingInterval[token][attendeeId])
		}

		Vue.set(state.speakingInterval[token], attendeeId, interval)
```

File: src/store/participantsStore.js [hunk 4]
```diff
@@ -433,9 +435,56 @@ const mutations = {
...
+	 */
+	updateTimeSpeaking(state, { token, attendeeId }) {
+		if (!state.speaking[token]?.[attendeeId]) {
+			return
+		}
+		if (!state.speaking[token][attendeeId].speaking) {
+			clearInterval(state.speakingInterval[token][attendeeId])
+			state.speakingInterval[token][attendeeId] = null
+			return
+		}
```
Comment (DorraJaouad) [inline src/store/participantsStore.js:481]:
```suggestion
		if (!state.speaking[token][attendeeId].lastTimestamp) {
			state.speaking[token][attendeeId].lastTimestamp = Date.now()
			return
		}
```

File: src/store/participantsStore.js [hunk 5]
```diff
@@ -1036,8 +1093,19 @@ const actions = {
...
+		if (speaking) {
+			interval = setInterval(() => {
+				context.commit('updateTimeSpeaking', { token, attendeeId })
+			}, 1000)
+		} else {
+			interval = null
+		}
+
+		context.commit('setSpeakingTime', { token, attendeeId, speaking })
+		context.commit('setSpeakingInterval', { token, attendeeId, interval })
```
Comment (DorraJaouad) [inline src/store/participantsStore.js:1108]:
```suggestion
		const interval= speaking ? setInterval(() => {
			context.commit('updateTimeSpeaking', { token, attendeeId })
		})
		: null
		context.commit('setSpeakingTime', { token, attendeeId, speaking })
		context.commit('setSpeakingInterval', { token, attendeeId, interval })
```

Comment (DorraJaouad) [reply src/store/participantsStore.js:1108]: So, the changes are meant to reduce the redundancy of clearing interval and make each mutation specified to a purpose/object.

File: src/components/RightSidebar/Participants/Participant.vue (hunk 1 above)
Comment (Antreesy) [reply src/components/RightSidebar/Participants/Participant.vue:872]: formattedTime expects `number`, and `this..isParticipantSpeking` could be `false`. But good catch anyway

**Review by DorraJaouad — COMMENTED** [review]
One last time 😬

File: src/store/participantsStore.js [hunk 6]
```diff
@@ -1038,6 +1079,26 @@ const actions = {
 
 	setSpeaking(context, { token, attendeeId, speaking }) {
 		context.commit('setSpeaking', { token, attendeeId, speaking })
+		context.commit('updateTimeSpeaking', { token, attendeeId, speaking, interval: false })
+
+		if (!context.state.speakingInterval) {
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply src/store/participantsStore.js:1084]: First signal can be `false`, but I'm agree, there's no need to start timers yet

[the remark this answers is not in the document]
Comment (DorraJaouad) [reply src/store/participantsStore.js:1084]: Yes, I noticed a signal is received from remote participants with speaking false then undefined at the beginning of the call.

**Pull request merged**